File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {    
    const visited = Array.from({length : words.length}, () => false)
    let min = Infinity
    
    function dfs(begin, answer){
        if(begin === target && min > answer){ 
            min = answer
            return
        }

        for(let i = 0; i < words.length; i++){
            let count = 0
            for(let j = 0; j < begin.length; j++){
                if(begin[j] !== words[i][j]){
                    count++
                }
            }
            if(count === 1 && !visited[i]){
                visited[i] = true
                dfs(words[i], answer + 1)
                visited[i] = false
            }
        }
    }
    
    dfs(begin, 0)
    return min === Infinity ? 0 : min
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def dfs(begin, target, words, count, visited) :
    if(begin == target) : 
        return count
    
    answer = len(words) + 1
    for i in range(len(words)) :
        c = 0
        if visited[i] == False :
            for l in range(len(words[i])) :
                if begin[l] != words[i][l] :
                    c += 1

            if c == 1 :
                visited[i] = True
                result = dfs(words[i], target, words, count + 1, visited)
                visited[i] = False
                if result != 0 : 
                    answer = min(answer, result)
    
    if answer == len(words) + 1 :
        return 0
    return answer

def solution(begin, target, words):
    visited = [False]  * len(words)
    
    return dfs(begin, target, words, 0, visited)